flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/LockRemover.java

```diff
@@ -0,0 +1,120 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/** Manages locks and collect {@link org.apache.flink.metrics.Metric} for the Maintenance Tasks. */
+@Internal
+public class LockRemover extends AbstractStreamOperator<MaintenanceResult>
+    implements OneInputStreamOperator<TaskResult, MaintenanceResult> {
+  private static final Logger LOG = LoggerFactory.getLogger(LockRemover.class);
+
+  private final TriggerLockFactory lockFactory;
+  private final List<String> taskNames;
+
+  private transient Map<Integer, Counter> successfulStreamResultCounterMap;
+  private transient Map<Integer, Counter> failedStreamResultCounterMap;
+  private transient Map<Integer, AtomicLong> lastRunLength;
+  private transient TriggerLockFactory.Lock lock;
+  private transient TriggerLockFactory.Lock recoveryLock;
+  private transient long lastProcessed = 0L;
+
+  public LockRemover(TriggerLockFactory lockFactory, List<String> taskNames) {
+    Preconditions.checkNotNull(lockFactory, "Lock factory should no be null");
+    Preconditions.checkArgument(
+        taskNames != null && !taskNames.isEmpty(), "Invalid task names: null or empty");
+
+    this.lockFactory = lockFactory;
+    this.taskNames = taskNames;
+  }
+
+  @Override
+  public void open() throws Exception {
+    super.open();
+    this.successfulStreamResultCounterMap = Maps.newHashMapWithExpectedSize(taskNames.size());
+    this.failedStreamResultCounterMap = Maps.newHashMapWithExpectedSize(taskNames.size());
+    this.lastRunLength = Maps.newHashMapWithExpectedSize(taskNames.size());
+    for (int i = 0; i < taskNames.size(); ++i) {
+      String name = taskNames.get(i);
+      successfulStreamResultCounterMap.put(
+          i,
+          getRuntimeContext()
+              .getMetricGroup()
+              .addGroup(TableMaintenanceMetrics.GROUP_KEY, name)
+              .counter(TableMaintenanceMetrics.SUCCESSFUL_STREAM_COUNTER));
+      failedStreamResultCounterMap.put(
+          i,
+          getRuntimeContext()
+              .getMetricGroup()
+              .addGroup(TableMaintenanceMetrics.GROUP_KEY, name)
+              .counter(TableMaintenanceMetrics.FAILED_STREAM_COUNTER));
+      AtomicLong length = new AtomicLong(0);
+      lastRunLength.put(i, length);
+      getRuntimeContext()
+          .getMetricGroup()
+          .addGroup(TableMaintenanceMetrics.GROUP_KEY, name)
+          .gauge(TableMaintenanceMetrics.LAST_RUN_LENGTH, length::get);
+    }
+
+    this.lock = lockFactory.createLock();
+    this.recoveryLock = lockFactory.createRecoveryLock();
+  }
+
+  @Override
+  public void processElement(StreamRecord<TaskResult> element) {
+    TaskResult value = element.getValue();
+    LOG.debug("TaskResult {} arrived", value);
+    long length = System.currentTimeMillis() - value.startEpoch();
+    output.collect(
+        new StreamRecord<>(
+            new MaintenanceResult(
+                value.startEpoch(), value.taskId(), length, value.success(), value.exceptions())));
```

Comment:
I'm open to discussion here.
Here are my thoughts:
- I prefer to have an output of the stream which could be logged or processed later. Having `Void` and changing later is backward incompatible. Having something for which we could add attributes could be compatible. That is why I have created the `MaintenanceResult`
- I would like to do the `duration` calculation (and metrics logging) in a single place, so I put it in here instead of putting it in every `Action`
- Inheritance for me means a bit more that some attributes are overlap. If we later need changes in one of the `MaintenanceResult` or the `TaskResult` it is easier to understand that they are independent.

Rejected alternatives:
- Keep only the `MaintenanceResult ` and fill the `duration` later - I hate "not filled" attributes which are later generated, because it is hard to follow when and why they got the value

Your thoughts?

Reply:
> I prefer to have an output of the stream which could be logged or processed later. Having Void and changing later is backward incompatible. Having something for which we could add attributes could be compatible. That is why I have created the MaintenanceResult

unless we have a clear goal of how to use the `MaintenanceResult`. Emitting it seems a bit pre-mature to me. I don't quite get the backward incompatible part. These classes like `LockRemover` is not user-facing classes.

Reply:
Removed